src/Control/Effect/Cull.hs

```diff
@@ -44,73 +44,45 @@ cull m = send (Cull m pure)
 -- | Run a 'Cull' effect. Branches outside of any 'cull' block will not be pruned.
 --
 --   prop> run (runNonDet (runCull (pure a <|> pure b))) == [a, b]
-runCull :: (Alternative m, Carrier sig m) => CullC m a -> m a
-runCull = (>>= runBranch (const empty)) . runReader False . runCullC
+runCull :: Alternative m => CullC m a -> m a
+runCull (CullC m) = runNonDetC (runReader False m) ((<|>) . pure) empty
 
-newtype CullC m a = CullC { runCullC :: ReaderC Bool m (Branch m () a) }
-  deriving (Functor)
-
-instance Alternative m => Applicative (CullC m) where
-  pure = CullC . pure . Pure
-  CullC f <*> CullC a = CullC (liftA2 (<*>) f a)
+newtype CullC m a = CullC { runCullC :: ReaderC Bool (NonDetC m) a }
```

Comment:
It’s really nice being able to express culling as the composition of nondeterminism and one bit of (local!) state ✨